demos/vc_issuer/src/main.rs

```diff
@@ -0,0 +1,521 @@
+        "name": "Bachelor of Engineering",
+        "institutionName": "DFINITY College of Engineering",
+      },
+    }))
+    .unwrap();
+
+    // Build credential using subject above and issuer.
+    CredentialBuilder::default()
+        .id(Url::parse("https://example.edu/credentials/3732").unwrap())
+        .issuer(Url::parse("https://example.edu").unwrap())
+        .type_("UniversityDegreeCredential")
+        .subject(subject)
+        .build()
+        .unwrap()
+}
+
+fn dfinity_employment_credential(subject_principal: Principal) -> Credential {
+    let subject: Subject = Subject::from_json_value(json!({
+      "id": did_for_principal(subject_principal),
+      "employee_of": {
+            "employerId" : "did:web:dfinity.org",
+            "employerName": "DFINITY Foundation",
+      },
+    }))
+    .unwrap();
+
+    // Build credential using subject above and issuer.
+    CredentialBuilder::default()
+        .id(Url::parse("https://employment.info/credentials/42").unwrap())
+        .issuer(Url::parse("https://employment.info").unwrap())
+        .type_("VerifiedEmployee")
+        .subject(subject)
+        .build()
+        .unwrap()
+}
+
+fn prepare_credential_payload(
+    req: &PrepareCredentialRequest,
+) -> Result<Credential, IssueCredentialError> {
+    match req.credential_spec.credential_name.as_str() {
+        "VerifiedEmployee" => {
+            EMPLOYEES.with_borrow(|employees| verify_authorized_principal(req, employees))?;
+            Ok(dfinity_employment_credential(req.signed_id_alias.id_alias))
+        }
+        "UniversityDegreeCredential" => {
+            GRADUATES.with_borrow(|graduates| verify_authorized_principal(req, graduates))?;
+            Ok(bachelor_degree_credential(req.signed_id_alias.id_alias))
+        }
+        other => Err(IssueCredentialError::UnsupportedCredentialSpec(format!(
+            "credential {} is not supported",
+            other
+        ))),
+    }
+}
+
+fn verify_authorized_principal(
+    req: &PrepareCredentialRequest,
+    authorized_principals: &HashSet<Principal>,
+) -> Result<(), IssueCredentialError> {
+    if authorized_principals.contains(&req.signed_id_alias.id_dapp) {
+        Ok(())
+    } else {
+        println!(
+            "*** principal {} it is not authorized for credential {}",
+            req.signed_id_alias.id_dapp.to_text(),
+            req.credential_spec.credential_name
+        );
+        Err(IssueCredentialError::UnauthorizedSubject(format!(
+            "unauthorized principal {}",
+            req.signed_id_alias.id_dapp.to_text()
+        )))
+    }
+}
+
+fn internal_error(msg: &str) -> IssueCredentialError {
+    IssueCredentialError::Internal(String::from(msg))
+}
+
+fn hash_bytes(value: impl AsRef<[u8]>) -> Hash {
+    let mut hasher = Sha256::new();
+    hasher.update(value.as_ref());
+    hasher.finalize().into()
+}
+
+// Order dependent: do not move above any function annotated with #[candid_method]!
+candid::export_service!();
```

Comment:
does this export the candid interface somehow? is this & the .did file kept in sync?

Reply:
Yes. The test below (`check_candid_interface_compatibility()`) verifies that the checked in `vc_issuer.did` is compatible with the did-file generated automatically.